### ISSUE TITLE
Rail tracks too easily griefable

### ISSUE THREAD

**Issue by Dodg-e** [jasonrohrer/OneLife#515]
Is it possible to keep track of who did an action on an object?

For example when someone uses an adze on a rail track is it possible to know who did it?

If it's the case it could be possible to make it so you need 2-3 different person using an adze on a rail track in 10 seconds for example for it to be dismantled

This could potentially be used for other stuff like walls, roads, oven, forge etc.

**Comment by jasonrohrer** [comment]
Hmm....

For adobe walls, the timing is such that it's hard to do by yourself.  This could be codified more strictly with an engine change (so that two people really did have to cooperate).

This is also what's achieved with the various elder mechanics (something gets stuck in one person's hand, requiring two people to do something together).

But as you're setting up a rail track, wouldn't it be a pain if you couldn't modify it?

Removing tracks is non-destructive (you still get the full track kit back).

The same is NOT true of the wall.  You put two adobe in and only get one back.

**Comment by jasonrohrer** [comment]
Hey, I realized that the "bend" action (smithing hammer to curve a straight track) is also griefable in just the same way.  If you have a working track, someone can mess it up with one hit of the smithing hammer.

Yet it would be ridiculous to require two people to bend a track, right?

Maybe not... I mean, tracks could be a necessarily 2-person operation.


Bending a track the wrong way is only SLIGHTLY less annoying than removing the track (because at least you don't have to pound stakes again to fix it).  But they're pretty much the same thing.

**Comment by Dodg-e** [comment]
But is it possible to remember who did an action?

This way one person could use the hammer and another person different would need to use it aswell in order to make the rail curve or dismantle rail

**Comment by jasonrohrer** [comment]
Well, we don't even need to change the code.

The hammer could "vibrate" (get stuck in the hand for X seconds), and the next hit to bend it would have to come in under X seconds for the track to bend.

However, the fundamental question is whether we WANT to require two people, working together with two hammers, to bend and otherwise modify tracks.

It's also a little messy content-wise, because tracks have so many cases.

What's the difference between griefing rails and griefing roads?  Both can currently be done solo, right?

**Comment by Dodg-e** [comment]
"vibrating" hammers and adzes would still require you to have two of these entities but ok.

Then the question become should roads be able to be as easily removed, i guess roads are sort of fine since the pickaxe enventually breaks and rebuilding road although a pain in the ass is not a big deal BUT in the case of railroads removing/bending rails and launching a cart means you destroy it and have to redo everything ALSO roads are easy to build it's one click with the stone on the stakes then place the flat rock, but the tracks have multiple directions,stopping points and even curves thanks to the last update so building them takes a lot more time (lot more clicks on the stakes), plus the pattern of the rail that is lost. (some railtracks look really nice)

**Comment by Dodg-e** [comment]
Also the vibrating thing is not twin,trip,quad proof that's why being able to know who did an action and if  they are related is essential imo

**Comment by jasonrohrer** [comment]
Yeah, but it's a pretty heavyweight solution to the problem....

Also, wouldn't rails then be used TO grief?

I mean, if they were hard to remove like that...

**Comment by Dodg-e** [comment]
It's a solution that could be applied to other objects like ancient waystones, bell towers, ancient walls etc so it wouldn't be only for rails.

That's a good point about using rails to grief but if it only requires 2-3 people to remove not a big issue especially since rails are limited.

Maybe leaders that have more than 2-3 followers could be able to remove rails alone?

Then griefers would beg for followers, they would look suspicious, players could catch up on that, still better than being able to directly remove everything

**Comment by jasonrohrer** [comment]
Yeah, leadership is an interesting thing to use for this purpose.

There could be a +leaderOnly tag that specified that only leaders could make some object.

**Comment by jasonrohrer** [comment]
Thought there's the problem of messaging.

For the other elder removal mechanics, it's messaged in the recipe hints.  You need an elder notice, etc, to remove this thing.  How do I get one of those?  Oh, an elder....

Same for the walls that require buckets of water to remove.

Invisible restrictions on action aren't so great.

**Comment by jasonrohrer** [comment]
Also, requiring two people to bend tracks seems like too big of a restriction.  It makes them inaccessible to solo players on low-pop.

And bending tracks is "just as bad" as removing them, in terms of disruption.

**Comment by Dodg-e** [comment]
I mean there is a lot of "hidden mechanics" atm

Leadership
Posse
Etc
Even emotes 

all arent specified anywhere except if you dig out old update posts or if some other players tells you "hey can you type "I JOIN YOU" or "I FOLLOW YOU" a little immersion breaking imo

And having this at the end of tutorial is not a good idea imo, because then you have to redo the whole tutorial if you forget something.

Having some sort of "help list/book" that regroups all these mechanics and different commands easily accessible would be nice, like a rulebook for the game but for mechanics.

"A leader with more than 3 followers can perform actions like removing public rails etc"

True for bending tracks, maybe the stakes should make the curve and not being able to bend them after with hammer.

**Comment by jasonrohrer** [comment]
What I mean is that most crafting things don't require hidden features to accomplish.

To tear down an adobe wall, you need to make it damp, and then it dries quickly.  This is shown in the hints that water is needed.  And it dries quickly, and it's clear what's going on, so you at least have a chance to figure out how to remove it with two people.

To have something that just doesn't work for an invisible reason is bad.

Age restrictions are kinda like this.... but babies can't do anything at first, so this is pretty clear to most people.

For things that only old people can do, I usually put ELDER ONLY in the name of the object.  I guess I could put LEADERS ONLY in the name of the object.  But that wouldn't specify how many followers would be needed.

Requiring more than 2 people to do something really puts a damper on low-pop play.

**Comment by Dodg-e** [comment]
I mean tools have visual text messaging (which shouldn't be imo)

Maybe when you try to remove a rail i tells you you dont have enough authority for it, not a fan of the idea tbh, better if it requires the effort of multiple people (makes more sense rails require multiple people effort).

Maybe to remove a rail you have to mark it first, like the elder notice to remove a fence, then it can be done in low pop too

except it would be a timed notice, you put the notice then wait x minutes if nobody removed it then it becomes approved and you can remove the rail

The reverse of turning something into "ancient" in a sense, could be done for walls, waystones, bells etc too btw.

10 hour notice to remove a wall that required 10 hours to become ancient seems fair.

**Comment by Dodg-e** [comment]
but then  you have to make a "notice" version of all the rails and that's not possible, nvm then...

unless the rail cart moving properties can be duplicated to the new "rail notice" object, then you only have to create the object and not the transition too.

**Comment by jasonrohrer** [comment]
Well, the notice rails could just block cart movement.

If someone cares about the rail and is using it, they will come along and remove the notice.

But it's still kind of a mess.... a griefer can place the notice and mess up your line.

Running around removing notices isn't interesting either.

**Comment by jasonrohrer** [comment]
Seems like the partially-dug rail that goes back to normal in 2 seconds, and a tool that gets stuck in your hand for 4 seconds, is the best option.  That won't be enough time to disrupt the train service (2 seconds of rail downtime here and there when a griefer fails to remove a rail).

But then we're back to bending rails, which could also require two people in a similar manner.... though bending a rail "live" is meant to allow people to re-rout trains between two connecting routes.

Though maybe no one is going to do that in practice?

**Comment by Dodg-e** [comment]
yeah but the tool stuck means low pop will get griefed with rails.

notice blocking cart is good imo, then you can just catch the griefer in action and curse+kill him.

**Comment by jasonrohrer** [comment]
Low pop can already get griefed with prop fences, walls, and other things that require two people to remove, right?

Twin up and fix that stuff if you need to on low-pop.

It also makes sense for many things to operate using the same pattern.  The 2-person-removal pattern is a good one.


The notices make sense for property fences because of the theme of homesteading.  Ancient way stones are also something that really makes sense for an elder to decide to remove.

**Comment by Tarr-OHOL** [comment]
Having someone rip up rails just as quickly as they can be put down is incredibly annoying to the point I expect anything I lay down to be ripped up for memes within 10 minutes of dying. 

Having the spring always break + one wheel means any time someone decides they want to create a toy track they grief the entire rail system meaning you need another spring + wheel. This also doesn't take into account any time the rail cart deletes the track under itself which then requires a while new blade + boards if you don't have spare track kits around. 


I don't know what the correct safeguard is for rails but if people are expected to make them for transporting stuff they can't be easily griefed else they'll just be regulated to making those dumb spinny tracks that do nothing at all.

**Comment by Dodg-e** [comment]
walls can be removed alone and fences decay.

needing multiple person to remove/bend rail is good but i thought low pop or maybe solo player server was the issue, alowing twins does have the issue of twin griefing though

**Comment by MonsWTF** [comment]
If I recall, Jason, you have setup code somewhere in OHOL where the games' objects can remember who interacted with them. (which is mainly coded for the purposes of feeding stuff to other people).

I think it might be worth expanding on the features of **Hierarchy, Investigation**, and **Law & Order** all in one go with the issues being presented here. 

**Investigation:** Add magnifying glass (or something) to the game. It becomes an unlearned Tool that, when used on an object + time, can "deduce" who last Used the object or triggered a Transition on it. In the case of the Track Rails being ripped up, someone could Magnifying Glass the rails OR the adze left at the "crime scene" to determine whodunnit.

**Law & Order:** Batons and Guillotines.

Baton: Non-lethal way of dispatching griefers/trolls in a village. Some semblance of bringing order along with the Hierarchy system that isn't just murder-knives. Person can then be rope'd to be dragged off to a Jail object (so that they can still be forcefed), or the Guillotine.

Guillotine: 3000-block area-ban on a player if ALL of the conditions for activating the Guillotine are met.
1) Person locked up in the stocks of the Guillotine (with the help of a Baton)
2) Group of X people all saying "Lynch Bob Lauren"
3) Someone high up in the Hierarchy is present (but doesn't have to be saying the Lynch phrase, just in case it's their head that's rolling)

Hierarchy restrictions: Allowing leaders of certain levels to have the ability to lock varying degrees of objects just like Property fences. Only certain people, like family, (or however they have the ability to restrict tools) being allowed to interact with them.

Lords: Can restrict fence gates
Barons: + can restrict carpentry tools/doors
Kings: + can restrict roads/rails
Emperors: + can restrict all other tools

**Comment by jasonrohrer** [comment]
Yeah, there is no tracking of who placed each object, currently.  I suppose that there could be, but it's a pretty deep and complicated change.  The magnifying glass is a nice mechanic, though.  Could also tell who left the tool here, or whatever else you want to punish (don't make a mess!).

Whodunnit also doesn't help to restore the lost time and work on a destroyed rail system, right?

I will fix carts to NOT lose anything when they go off the rails (no broken spring or lost wheels).

**Comment by Dodg-e** [comment]
it's not only about the cart, it's all the work that you did to make the rail system that can just be destroyed in a few seconds

**Comment by MonsWTF** [comment]
The only thing that immediately comes to mind as far as fixing the rail system being easily griefable is to have a mechanic where each rail type can be Transition'd into a permanent rail, either over time or with some elder involvement.

But then you're talking about adding a duplicate set of rails for each of the existing ones, and re-creating alllll those categories it took to build it.

So that's why I haven't really mentioned it.

If the Editor was able to make it easy to clone over all the transitions, categories, etc. into a duplicate set of "permanent" rails, and you kept the non-permanent rails separate from the permanent rails, then technically you could have the non-grief'able rails. But then you're adding a permanent ground tile object to the game, that if not designed properly can become even more of a hassle.

**Comment by jasonrohrer** [comment]
Yes, dodge, I'm also making track removal require two people.

And track-bending will only be do-able for a short window (like 30 seconds), instead of forever.

Thus, tracks become single-player permanent after they are built, and require two people to modify or remove.

This will use the "tool stuck in hand" and "two hits required" pattern.

On top of that, a cart going off rails won't destroy the spring.
